### PR TITLE
#5030 Bug fix  "Author field shows HTML tags when template used"

### LINKED ISSUE #5030: Author field shows HTML tags when template used
### Summary

When viewing my recent uploads in the app (using Beta version=4.0.3~23c474b06), the attribution includes visible HTML markup and incorrect text. This is apparently caused when `{{Creator:Andy Mabbett}}` is present on the Commons file page, as it includes the description from the associated Wikidata item. For items with the author in plain text, the issue does not occur.

### Steps to reproduce

As above

### Expected behaviour

Correct attribution should be shown.

### Actual behaviour

As above

### Device name

_No response_

### Android version

Android 11

### Commons app version

4.0.3~23c474b06

### Device logs

_No response_

### Screen-shots

https://commons.wikimedia.org/wiki/File:Commons_app_beta_screenshot_showing_attrbution_text_errors_-_2022-08-07.jpg  

### Would you like to work on the issue?

Prefer not

### PR DESCRIPTION
**Pull request by shankarpriyank**
Added a new function getDisplayAuthor in Media.kt that parses the author string with Regex, also added some tests for it

**Description (required)**
Fixes #5030

Just added a few tests to the changes made in #5069

### CODE REVIEW

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 1]
```diff
@@ -114,6 +114,30 @@ class Media constructor(
             else
                 ""
 
+    /**
+     * Gets media display title
+     * @return Media title
+     */
+
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:121]: Remove extra line between documentation and function

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 2]
```diff
@@ -114,6 +114,30 @@ class Media constructor(
...
+    fun getDisplayAuthor(): String{
+        if (author == null)
+            return ""
+
+        val completeHtmlTagCheck = "<.*>(.*)<.*>".toRegex() // Detect if complete html tags present
+        var completeCheckResult = completeHtmlTagCheck.findAll(author!!) // Find matching cases, return text between tag
+        if(completeCheckResult.count()==1) return completeCheckResult.toList().get(0).groupValues.get(1) // Get matching result and return text between tag - text between tag is inside a capture group
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:128]: Please use these spacing conventions: https://kotlinlang.org/docs/coding-conventions.html#horizontal-whitespace

Also, better split this into 3 lines:
```
// comment
if ...
    return ...

```
with a new line before and after.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 3]
```diff
@@ -114,6 +114,30 @@ class Media constructor(
...
+        else if (completeCheckResult.count()>1) return "" // If there are multiple cases we return nothing --> can multiple cases occur?
+
+        val hrefTagCheck = "<.*>(.*)".toRegex() // Detect if half html tags present -> <a href= XXX> case
+        var hrefTagCheckResult = hrefTagCheck.findAll(author!!) // Find matching cases, return text between tag
+        if(hrefTagCheckResult.count()==1) return hrefTagCheckResult.toList().get(0).groupValues.get(1) // Get matching result and return text between tag
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:133]: new line before, spacing, split each of these two lines into two.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 4]
```diff
@@ -114,6 +114,30 @@ class Media constructor(
...
+        else if (hrefTagCheckResult.count()>1) return "" // If there are multiple cases we return nothing --> can multiple cases occur?
+
+        return author!! // If this is reached then html tags are not present, return author text as is
+
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:137]: no need for a new line here.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 5]
```diff
@@ -114,6 +114,30 @@ class Media constructor(
...
+    }
+
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:139]: Only one new line is enough.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
A few stylistic changes requested.
Thanks for the unit tests!
I will now test this change for a while with real-world users.

**Comment by shankarpriyank** [comment]
@nicolas-raoul I have made the changes as requested, also please approve the ci check, there were 2 failing tests earlier I have fixed them.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 6]
```diff
@@ -114,6 +114,39 @@ class Media constructor(
...
+    fun getDisplayAuthor(): String{
+
+        if (author == null)
+            return ""
+
+        val completeHtmlTagCheck = "<.*>(.*)<.*>".toRegex() // Detect if complete html tags present
+        var completeCheckResult = completeHtmlTagCheck.findAll(author!!) // Find matching cases, return text between tag
+
+        // Get matching result and return text between tag - text between tag is inside a capture group
+        if(completeCheckResult.count()==1)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:130]: Spaces before and after `==` here and below.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 7]
```diff
@@ -114,6 +114,39 @@ class Media constructor(
...
+            return completeCheckResult.toList().get(0).groupValues.get(1)
+
+        // If there are multiple cases we return nothing --> can multiple cases occur?
+        else if (completeCheckResult.count()>1)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:134]: Spaces before and after `>`.

File: app/src/main/java/fr/free/nrw/commons/Media.kt (hunk 6 above)
Comment (nicolas-raoul) [reply app/src/main/java/fr/free/nrw/commons/Media.kt:130]: Space after `if` here and below.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Thanks! Just a few spacing issues left.

File: app/src/main/java/fr/free/nrw/commons/Media.kt [hunk 8]
```diff
@@ -114,6 +114,39 @@ class Media constructor(
             else
                 ""
 
+    /**
+     * Gets media display title
+     * @return Media title
+     */
+    fun getDisplayAuthor(): String{
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/Media.kt:121]: Space before `{`.

**Comment by shankarpriyank** [comment]
@nicolas-raoul Sorry for the iteration, I have made the changes as requested.
I also have a suggestion, maybe we can a linter to the project and use a pre commit hook to run a basic linting command and a linting check. Then most of the styling issues will be automatically fixed.

**Review by nicolas-raoul — COMMENTED** [review]
Looks great now!

Would you mind checking whether anything can be done for cases such as the one I described at https://github.com/commons-app/apps-android-commons/issues/5030#issuecomment-1482386712? Why does the output becomes empty? If this looks like a different problem, feel free to create a separate issue. Thanks a lot!

**Comment by shankarpriyank** [comment]
That seems like a different issue to be right now, but I think we can investigate a lil more before jumping to conclusions([I have asked for some info in the issue](https://github.com/commons-app/apps-android-commons/issues/5030#issuecomment-1483291921)).

Also the pr still shows _changes requested._

**Comment by nicolas-raoul** [comment]
The pull request still shows changes requested even though I marked all issues as resolved, you can disregard this status.

In the case of https://commons.wikimedia.org/wiki/File:Mastl%C3%A9_Odles_Stevia.jpg what is the string that `getDisplayAuthor()` receives as input? Please paste the whole input string here, thanks!

**Comment by shankarpriyank** [comment]
@nicolas-raoul 
`<tbody><tr class="plainlinks"><td><a href="//commons.wikimedia.org/wiki/File:Emblem-camera.svg" class="image">`

This is what we get as the author(the input in getDisplayAuthor). As Ritika said, this indeed is the icon of a camera. No useful information is present in the author field

**Comment by nicolas-raoul** [comment]
Strange, it seems to be cut. The full HTML should be much larger, containing all of the other parts of the template such as the email image and the "Wolfgang Moroder" text.

Would you mind checking whether our app is cutting this HTML, or whether the API gives us only this truncated version? If the latter, we need to find a way to get the whole HTML, as not showing the author name might be a copyright license violation.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Sorry, we should first get the author name from Structured Data, as discussed recently at https://github.com/commons-app/apps-android-commons/issues/5030#issuecomment-1483789016. Thank you for the several great improvements to this pull request!

**Comment by shankarpriyank** [comment]
I will have a look, although this may take some time though as this seems a bit new to me

**Comment by domdomegg** [comment]
Closing this as now stale

**Pull request closed without merging**